frontend/src/metabase/visualizations/types/visualization.ts

```diff
@@ -176,24 +193,59 @@ export type VisualizationSettingDefinition<TValue, TProps = void> = {
   getMarginBottom?: (series: Series, settings: VisualizationSettings) => string;
   persistDefault?: boolean;
   inline?: boolean;
-  props?: TProps;
-  getProps?: (
-    series: Series,
-    vizSettings: VisualizationSettings,
-    onChange: (value: TValue) => void,
-    extra: unknown,
-    onChangeSettings: (value: Record<string, any>) => void,
-  ) => TProps;
   readDependencies?: string[];
   writeDependencies?: string[];
   eraseDependencies?: string[];
-  // is the setting visible in the dashboard card viz settings
   dashboard?: boolean;
   useRawSeries?: boolean;
+  hidden?: boolean;
+  placeholder?: string;
+  index?: number;
 };
 
+export type WidgetComponentDef<Widget, P, TValue> = {
+  widget?: Widget;
+  props?: P;
+  getProps?: (
+    series: Series,
+    vizSettings: VisualizationSettings,
+    onChange: (value: TValue | null) => void,
+    extra: SettingDefinitionExtra,
+    onChangeSettings: (value: VisualizationSettings) => void,
+  ) => P;
+};
```

Comment:
Could we use the same naming convention for all generic type parameters? imo, it makes it slightly easier to read.

```suggestion
export type WidgetComponentDef<Widget, Props, Value> = {
  widget?: Widget;
  props?: Props;
  getProps?: (
    series: Series,
    vizSettings: VisualizationSettings,
    onChange: (value: Value | null) => void,
    extra: SettingDefinitionExtra,
    onChangeSettings: (value: VisualizationSettings) => void,
  ) => Props;
};
```